src/GraphQL/Resolver.hs

```diff
@@ -227,63 +227,76 @@ instance forall m ksN enum. (Applicative m, API.GraphQLEnum enum) => HasGraph m
--- name doesn't match.
-data NamedValueResolver m = NamedValueResolver Name (m (Result Value))
 
 -- Iterate through handlers (zipped together with their type
 -- definition) and execute handler if the name matches.
+
+-- TODO: A parametrized `Result` is really not a good way to handle
+-- the "result" for resolveField, but not sure what to use either. I
+-- liked the tuple we had before more because it didn't imply any
+-- other structure or meaning. Maybe we can jsut create a new datatype.
 type ResolveFieldResult = Result (Maybe GValue.ObjectField)
 
-resolveField :: forall resolverType (m :: Type -> Type). (BuildFieldResolver m resolverType, Monad m)
-  => FieldHandler m resolverType -> m ResolveFieldResult -> Field Value -> m ResolveFieldResult
+-- Extract field name from an argument type. TODO: ideally we'd run
+-- this directly on the "a :> b" argument structure, but that requires
+-- passing in the plain argument structure type into resolveField or
+-- resolving "name" in the buildFieldResolver. Both options duplicate
+-- code somwehere else.
+type family FieldName (a :: Type) = (r :: Symbol) where
+  FieldName (JustHandler (API.Field name t)) = name
+  FieldName (PlainArgument a f) = FieldName f
+  FieldName (EnumArgument a f) = FieldName f
+  FieldName x = TypeError ('Text "Unexpected branch in FieldName type family. Please file a bug!" ':<>: 'ShowType x)
+
+resolveField :: forall dispatchType (m :: Type -> Type).
+  (BuildFieldResolver m dispatchType, Monad m, KnownSymbol (FieldName dispatchType))
+  => FieldHandler m dispatchType -> m ResolveFieldResult -> Field Value -> m ResolveFieldResult
 resolveField handler nextHandler field =
-  case buildFieldResolver @m @resolverType handler field of
-    -- TODO the fact that this doesn't fit together nicely makes me think that ObjectField is not a good idea)
-    Left err -> pure (Result [err] (Just (GValue.ObjectField queryFieldName GValue.ValueNull)))
-    Right (NamedValueResolver name' resolver) -> runResolver name' resolver
+  -- check name before
+  case nameFromSymbol @(FieldName dispatchType) of
+    Left err -> pure (Result [SchemaError err] (Just (GValue.ObjectField queryFieldName GValue.ValueNull)))
+    Right name' -> runResolver name'
   where
-    runResolver :: Name -> m (Result Value) -> m ResolveFieldResult
-    runResolver name' resolver
-      | queryFieldName == name' = do
-          Result errs value <- resolver
-          pure (Result errs (Just (GValue.ObjectField queryFieldName value)))
+    runResolver :: Name -> m ResolveFieldResult
+    runResolver name'
+      | queryFieldName == name' =
+          case buildFieldResolver @m @dispatchType handler field of
+            Left err -> pure (Result [err] (Just (GValue.ObjectField queryFieldName GValue.ValueNull)))
+            Right resolver -> do
+              Result errs value <- resolver
+              pure (Result errs (Just (GValue.ObjectField queryFieldName value)))
```

Comment:
I'm not entirely sure what you mean, sorry! Would you mind folding the change into your next branch?